backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/ReturnsInsertionLowering.kt

```diff
@@ -32,14 +31,25 @@ internal class ReturnsInsertionLowering(val context: Context) : FileLoweringPass
             override fun visitFunction(declaration: IrFunction) {
                 declaration.acceptChildrenVoid(this)
 
-                val body = declaration.body
-                if ((declaration is IrConstructor || declaration.returnType.classifierOrNull == symbols.unit) && body != null) {
+                val body = declaration.body ?: return
+                if (declaration is IrConstructor || declaration.returnType == context.irBuiltIns.unitType) {
                     val irBuilder = context.createIrBuilder(declaration.symbol, declaration.endOffset, declaration.endOffset)
                     irBuilder.run {
                          (body as IrBlockBody).statements += irReturn(irGetObject(symbols.unit))
                     }
                 }
             }
+
+            override fun visitBlock(expression: IrBlock) {
+                expression.acceptChildrenVoid(this)
+                if (expression !is IrReturnableBlock) return
+                if (expression.inlineFunctionSymbol?.owner?.returnType == context.irBuiltIns.unitType) {
```

Comment:
Which inline function has `Unit` return type in the test case?

Reply:
The lambda being passed to `let` function.